Application parameter for :download-open/:prompt-open-download and tests

Kingdread opened the pull request:
Follow-up for #1730. Fixes #1744, #1211 and adds tests for #1728 and #39.

Tests for #1725 and #1726 are still missing, I have to see how I get the suggested filename long enough :wink:

The syntax is `:download-open echo {}` where `{}` is replaced by the filename. The same for `:prompt-open-download echo {}`, though you need to bind it to a key, so you probably only use it if you have a special application.

The-Compiler commented:
> Tests for #1725 and #1726 are still missing, I have to see how I get the suggested filename long enough :wink:

There's a httpbin instance running, so you could probably do something with http://httpbin.org/response-headers?Content-Disposition=download;%20filename=some-very-long-name

Kingdread commented:
I've worked on the remaining tests, but locally I get a `QProcess: Destroyed while process ("/home/daniel/Code/qutebrowser/.tox/py35/bin/python") is still running.` (in `test_invocations.py`), even though I use `start_detached`, which in turn uses `QProcess::startDetached`, which according to the [Qt docs](http://doc.qt.io/qt-5/qprocess.html#startDetached):

> [...] If the calling process exits, the detached process will continue to run unaffected.

I'll do some more investigative work tomorrow.

The-Compiler commented:
I've seen that message being printed before (only occasionally), so I'm not sure if it's because of your code. If you see it on `master` too and it doesn't fail anything, feel free to ignore it.

The-Compiler commented on tests/end2end/data/downloads/issue1725.html:
```diff
@@ -0,0 +1,13 @@
+    <body>
+        <p>Using <code>:prompt-open-download</code> with a file that has a loooooong filename</p>
+        <p>
+            <a href="/response-headers?Content-Disposition=download;%20filename%2A%3DUTF-8%27%27Lookie%2520here%252C%2520your%2520compiler%2520does%2520some%2520absolutely%2520insane%2520things%2520with%2520the%2520spilling%252C%2520including%2520spilling%2520a%2520%252Aconstant%252A.%2520For%2520chrissake%252C%2520that%2520compiler%2520shouldn%2527t%2520have%2520been%2520allowed%2520to%2520gradu ...
```
...what. :laughing:

Kingdread replied:
It's a quote by [Linus Torvalds](https://en.wikiquote.org/wiki/Linus_Torvalds#2014) that was long enough to trigger the error, doubly percent-encoded to get all special characters right :wink:

I actually didn't think about your nickname when I chose it, so that's an unfortunate coincdence. I'm happy to change it if you don't feel like being called "not being allowed to graduate kindergarten", sorry about that...

The-Compiler replied:
Ah! :laughing: No worries, but I wouldn't mind some different Torvalds quote. Why not this one? :wink:

> I've actually felt slightly uncomfortable at TED for the last two days, because there's a lot of vision going on, right? And I am not a visionary. I do not have a five-year plan. I'm an engineer. And I think it's really -- I mean -- I'm perfectly happy with all the people who are walking around and just staring at the clouds and looking at the stars and saying, "I want to go there." But I'm looking at the ground, and I want to fix the pothole that's right in front of me before I fall in. This is the kind of person I am.

The-Compiler commented:
LGTM so far, let me know when the rest is ready :+1:

Kingdread commented:
I'm not seeing the QProcess message when removing my test code, so it seems to be triggered by my additions. It also doesn't appear on the archlinux travis build (though I'm also running arch locally), so I'm thinking it is more of a Qt thing...

Kingdread commented:
I've changed the text and fixed a test that was passing regardless of the filename, and I've tested that each test fails when I remove the corresponding fix, so it should be fine.

I'd say this is ready now depending on the QProcess message thing.

Kingdread commented on tests/end2end/data/downloads/issue1725.html:
```diff
@@ -0,0 +1,13 @@
+    <body>
+        <p>Using <code>:prompt-open-download</code> with a file that has a loooooong filename</p>
+        <p>
+        <a href="/response-headers?Content-Disposition=download;%20filename%2A%3DUTF-8%27%27I%2527ve%2520actually%2520felt%2520slightly%2520uncomfortable%2520at%2520TED%2520for%2520the%2520last%2520two%2520days%252C%2520because%2520there%2527s%2520a%2520lot%2520of%2520vision%2520going%2520on%252C%2520right%253F%2520And%2520I%2520am%2520not%2520a%2520visionary.%2520I%2520do%2520not%2520have%2520a% ...
```
Seems like GitHub doesn't want to syntax-highlight this line... :stuck_out_tongue_winking_eye:

The-Compiler commented:
I just noticed there are some more small things:

- I wonder, should we really have the `{}` or just pass the filename as first argument? I already caught myself trying to do `:download-open gvim` and wondering why it didn't open the file...
- You sometimes have `{}` without backticks and sometimes with - I think there should be backticks everywhere
- You use "default program", "default application" and simply "default" in the docs - I'd propose to always go for "default application"

Kingdread commented:
I think we could do what @NoctuaNivalis proposed [here](https://github.com/The-Compiler/qutebrowser/issues/1211#issuecomment-168810542) and append the filename if no `{}` was found. Then `:download-open gvim` should work.

I will fix the other two things.

The-Compiler commented:
Sounds good!

Kingdread commented on qutebrowser/utils/usertypes.py:
```diff
@@ -297,11 +297,16 @@ def __init__(self, fileobj):
 
-    def __init__(self):
+    Attributes:
+        cmdline: The command to use as string. A {} is expanded to the
```
I've not added backticks here because this is no command, the backticks were mainly for asciidoc. Should I still add them?

The-Compiler replied:
Please do - at some point I want to have autogenerated (Sphinx) documentation for all docstrings, as developer/API docs.

Kingdread commented:
Fixed, `:download-open gvim` should now work!

The-Compiler commented:
Can you please also point that out in the respective docstrings?

Kingdread commented:
Sure, did it

The-Compiler commented:
Thank you!
